File: bindings/java/src/main/java/com/ainos/sdk/models/ModelLoadOptions.java
package com.ainos.sdk.models;

import java.util.Objects;

/**
 * Options for loading a model into the Ainos daemon.
 * <p>
 * This class is immutable and should be constructed using the {@link Builder}.
 *
 * <pre>{@code
 * ModelLoadOptions opts = ModelLoadOptions.builder()
 *     .architecture("phi3")
 *     .gpuLayerCount(32)
 *     .contextSize(4096)
 *     .build();
 * }</pre>
 */
public final class ModelLoadOptions {

    private final String architecture;
    private final Integer gpuLayerCount;
    private final Integer contextSize;
    private final Boolean useMmap;
    private final Integer threads;
    private final String engineType;

    private ModelLoadOptions(Builder builder) {
        this.architecture = builder.architecture;
        this.gpuLayerCount = builder.gpuLayerCount;
        this.contextSize = builder.contextSize;
        this.useMmap = builder.useMmap;
        this.threads = builder.threads;
        this.engineType = builder.engineType;
    }

    /**
     * Creates a new {@link Builder} for constructing {@code ModelLoadOptions}.
     *
     * @return a new builder instance
     */
    public static Builder builder() {
        return new Builder();
    }

    /** Returns the model architecture hint, if set. */
    public java.util.Optional<String> getArchitecture() {
        return java.util.Optional.ofNullable(architecture);
    }

    /** Returns the number of GPU layers to offload, if set. */
    public java.util.Optional<Integer> getGpuLayerCount() {
        return java.util.Optional.ofNullable(gpuLayerCount);
    }

    /** Returns the context size in tokens, if set. */
    public java.util.Optional<Integer> getContextSize() {
        return java.util.Optional.ofNullable(contextSize);
    }

    /** Returns whether memory-mapped I/O should be used, if set. */
    public java.util.Optional<Boolean> getUseMmap() {
        return java.util.Optional.ofNullable(useMmap);
    }

    /** Returns the number of CPU threads for inference, if set. */
    public java.util.Optional<Integer> getThreads() {
        return java.util.Optional.ofNullable(threads);
    }

    /** Returns the engine type override, if set (e.g. {@code "ggml"}, {@code "onnx"}). */
    public java.util.Optional<String> getEngineType() {
        return java.util.Optional.ofNullable(engineType);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ModelLoadOptions)) return false;
        ModelLoadOptions that = (ModelLoadOptions) o;
        return Objects.equals(architecture, that.architecture)
                && Objects.equals(gpuLayerCount, that.gpuLayerCount)
                && Objects.equals(contextSize, that.contextSize)
                && Objects.equals(useMmap, that.useMmap)
                && Objects.equals(threads, that.threads)
                && Objects.equals(engineType, that.engineType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(architecture, gpuLayerCount, contextSize, useMmap, threads, engineType);
    }

    @Override
    public String toString() {
        return "ModelLoadOptions{"
                + "architecture='" + architecture + '\''
                + ", gpuLayerCount=" + gpuLayerCount
                + ", contextSize=" + contextSize
                + ", useMmap=" + useMmap
                + ", threads=" + threads
                + ", engineType='" + engineType + '\''
                + '}';
    }

    // -----------------------------------------------------------------------
    // Builder
    // -----------------------------------------------------------------------

    /**
     * Builder for {@link ModelLoadOptions}.
     */
    public static final class Builder {
        private String architecture;
        private Integer gpuLayerCount;
        private Integer contextSize;
        private Boolean useMmap;
        private Integer threads;
        private String engineType;

        private Builder() {
        }

        /** Sets the model architecture hint (e.g. {@code "auto"}, {@code "phi3"}, {@code "llama"}). */
        public Builder architecture(String architecture) {
            this.architecture = architecture;
            return this;
        }

        /** Sets the number of GPU layers to offload. */
        public Builder gpuLayerCount(int gpuLayerCount) {
            this.gpuLayerCount = gpuLayerCount;
            return this;
        }

        /** Sets the context size in tokens. */
        public Builder contextSize(int contextSize) {
            this.contextSize = contextSize;
            return this;
        }

        /** Sets whether to use memory-mapped I/O for model loading. */
        public Builder useMmap(boolean useMmap) {
            this.useMmap = useMmap;
            return this;
        }

        /** Sets the number of CPU threads for inference. */
        public Builder threads(int threads) {
            this.threads = threads;
            return this;
        }

        /** Sets the engine type override (e.g. {@code "ggml"}, {@code "onnx"}). */
        public Builder engineType(String engineType) {
            this.engineType = engineType;
            return this;
        }

        /** Builds the {@link ModelLoadOptions}. */
        public ModelLoadOptions build() {
            return new ModelLoadOptions(this);
        }
    }
}

File: bindings/csharp/AinosSdk/Models/ModelLoadOptions.cs
using System.Text.Json.Serialization;

namespace AinosSdk.Models;

/// <summary>
/// Optional parameters for loading a model.
/// </summary>
public class ModelLoadOptions
{
    /// <summary>
    /// Whether to skip loading if the model is already loaded.
    /// </summary>
    [JsonPropertyName("skip_if_loaded")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public bool SkipIfLoaded { get; init; }

    /// <summary>
    /// Override the model architecture (e.g. "auto", "phi3", "llama").
    /// </summary>
    [JsonPropertyName("architecture")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Architecture { get; init; }

    /// <summary>
    /// Number of GPU layers to offload (0 = CPU only).
    /// </summary>
    [JsonPropertyName("gpu_layers")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? GpuLayers { get; init; }

    /// <summary>
    /// Context size in tokens.
    /// </summary>
    [JsonPropertyName("context_size")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? ContextSize { get; init; }

    /// <summary>
    /// Creates a default <see cref="ModelLoadOptions"/>.
    /// </summary>
    public static ModelLoadOptions Default => new();

    /// <summary>
    /// Creates a new <see cref="Builder"/> for constructing <see cref="ModelLoadOptions"/>.
    /// </summary>
    public static Builder CreateBuilder() => new();

    /// <summary>
    /// Fluent builder for <see cref="ModelLoadOptions"/>.
    /// </summary>
    public class Builder
    {
        private bool _skipIfLoaded;
        private string? _architecture;
        private int? _gpuLayers;
        private int? _contextSize;

        internal Builder() { }

        /// <summary>Sets whether to skip loading if already loaded.</summary>
        public Builder WithSkipIfLoaded(bool skip = true) { _skipIfLoaded = skip; return this; }

        /// <summary>Sets the model architecture override.</summary>
        public Builder WithArchitecture(string architecture) { _architecture = architecture; return this; }

        /// <summary>Sets the number of GPU layers.</summary>
        public Builder WithGpuLayers(int gpuLayers) { _gpuLayers = gpuLayers; return this; }

        /// <summary>Sets the context size in tokens.</summary>
        public Builder WithContextSize(int contextSize) { _contextSize = contextSize; return this; }

        /// <summary>Builds the <see cref="ModelLoadOptions"/>.</summary>
        public ModelLoadOptions Build() => new()
        {
            SkipIfLoaded = _skipIfLoaded,
            Architecture = _architecture,
            GpuLayers = _gpuLayers,
            ContextSize = _contextSize,
        };
    }
}